scripts/build.js

```diff
@@ -89,13 +93,95 @@ function checkChecksum(url, local_file) {
     })
 }
 
+/**
+ * Downloads an NPM package for use dynamically - so it will only be loaded and used for this single script.
+ * What it does under the hood is check for the npm package in the node modules, then in the npm list, and if
+ * it does not find it in either location, it will download the package at that point, adding it as a dev-dependency.
+ *
+ * It it downloads it dynamically, then it will return true. This is so you can delete the package once you are done,
+ * so it doesn't leave a zombie package in your node_modules. To remove the package, call npmDeleteRuntimePackage
+ *
+ * @param {*} package_name The name of the package you want to download (example: 'cmake-js')
+ * @param {*} package_version The version of the package to download - leave blank for latest. (example: '6.3.2')
+ * @returns True if the package was downloaded dynamically, otherwise false.
+ */
+function npmDownloadAndInstallRuntimePackage(package_name, package_version=null) {
+    console.log("Looking for " + package_name + " as a dependency...");
+
+    // Do we have it in node modules? If so, then use that!
+    try {
+        if (fs.existsSync("./node_modules/" + package_name)) {
+            console.log("Found " + package_name + " in node_modules!");
+            return false;
+        }
+    } catch (error) {}
+
+    // Do we have it in our node list? If so, then use that!
+    try {
+        var list_output = child_process.execSync("npm list " + package_name, {encoding: "utf8"});
+        if (list_output.indexOf(package_name) !== -1) {
+            console.log("Found " + package_name + " in npm list!");
+            return false;
+        }
+    } catch (error) {}
+
+    // If it is not found in either, then download it into our node_modules
+    try {
+        console.log("Could not find " + package_name);
+        console.log("Downloading " + package_name + " from npm for build...");
+        // Try to intall the given package and ONLY the given package. Will throw an exception if there is an error.
+        if (package_version != null) {
+            child_process.execSync("npm install --no-package-lock --save-dev --ignore-scripts " + package_name + "@" + package_version);
```

Comment:
This seems really sketchy

I realize the whole premise of downloading code and compiling it as part of the install process is already a tough pill to swallow, but is it going to far to install additional NPM dependencies that aren't declared in `package.json`?

Let's talk with someone that has deeper familiarity with what is considered acceptable in Javascript-Land

But if this is acceptable, we could do it with `cmake-js` as well 🤷‍♀️

Reply:
I will leave it for now until and see if I can get more feedback to make sure the change is okay with those experienced in Javascript 👍

Reply:
For good or bad, this doesn't faze me as a JS developer. 

I suggest doing this in a subfolder workspace such as `your_package_location/scripts/`.
You can leave a dummy `package.json` with `private:true` which will allow temporary modules to be installed there if they are not part of the user's workspace. 

That way, the workflow could look like this:

- user runs `npm install aws-crt`
- the `aws-crt` package is downloaded to their node_modules (root)
    - `aws-crt` postinstall tries to use the root `cmake-js` et. al. runtime packages
    - if not present or of the wrong version, download them into a subfolder of `aws-crt` so as not to affect the user's root workspace node_modules / package.json / package-lock.json etc.
    - if not able to download, let the user know they will need to install certain dev dependencies and re-run the `aws-crt` install script.

This would avoid bringing non-runtime packages into both the user's development workspace (root node_modules) as well as the user's potential runtime distribution (package.json->dependencies).